File: cpp/leetcode/229.cpp
class Solution {
public:
    vector<int> majorityElement(vector<int>& nums) {
        int cnts1 = 0, cnts2 = 0;
        int nums1 = 0, nums2 = 1; // no matter the exactly value, just different
        for(auto num : nums){
            if(num == nums1){
                cnts1 ++;
            } else if (num == nums2){
                cnts2 ++;
            } else if (cnts1 == 0) {
                nums1 = num;
                cnts1 ++;
            } else if (cnts2 == 0) {
                nums2 = num;
                cnts2 ++;
            } else {
                cnts1 --;
                cnts2 --;
            }
        }
        cnts1 = 0; cnts2 = 0;
        for(auto num : nums){
            if(num == nums1) cnts1 ++;
            else if(num == nums2) cnts2++;
        }
        vector<int> res;
        if(cnts1 > nums.size() / 3)  res.push_back(nums1);
        if(cnts2 > nums.size() / 3)  res.push_back(nums2);
        return res;
    }
};

File: python/leetcode/229.py
class Solution:
    def majorityElement(self, nums):
        if not nums: return []
        cnts1, cnts2, num1, num2 = 0, 0, 0, 1
        for n in nums:
            if n == num1:
                cnts1 += 1
            elif n == num2:
                cnts2 += 1
            elif cnts1 == 0:
                num1, cnts1 = n, 1
            elif cnts2 == 0:
                num2, cnts2 = n, 1
            else:
                cnts1, cnts2 = cnts1 - 1, cnts2 - 1
        return [n for n in (num1, num2) if nums.count(n) > len(nums)//3]